message Function {
    required string name = 1;
    required string return_type = 2;
    repeated Parameter parameter = 3;

    message Parameter {
        required string name = 1;
        required string type = 2;
    }
}

message ObjCMethod {
    required Function function = 1;
    required bool class_method = 2;
}

message ObjCClass {
    required string name = 1;
    optional string base_class = 2;
    repeated string protocol = 3;
    repeated ObjCMethod method = 4;
}

message ObjCProtocol {
    required string name = 1;
    repeated string base_protocol = 2;
    repeated ObjCMethod method = 3;
}

message ObjCCategory {
    required string class_name = 1;
    required string category_name = 2;
    repeated string base_protocol = 3;
    repeated ObjCMethod method = 4;
}

message TranslationUnit {
    repeated ObjCClass class = 1;
    repeated ObjCProtocol protocol = 2;
    repeated ObjCCategory category = 3;
}
